Panic when pasting large text twice

If I copy `core-lib/view.rs` and paste it twice then xi panics:

```
thread '<xi-syntect-plugin> core host thread' panicked at 'assertion failed: self.start >= amount', rope/src/interval.rs:135:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: xi_rope::interval::Interval::translate_neg
   7: xi_core_lib::linewrap::Lines::patchup_tasks
   8: xi_core_lib::linewrap::Lines::after_edit
   9: xi_core_lib::view::View::after_edit
  10: xi_core_lib::event_context::EventContext::update_views::{{closure}}
  11: core::iter::iterator::Iterator::for_each::{{closure}}
  12: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  13: core::iter::iterator::Iterator::fold::{{closure}}
  14: core::iter::iterator::Iterator::try_fold
  15: core::iter::iterator::Iterator::fold
  16: <core::iter::Chain<A, B> as core::iter::iterator::Iterator>::fold
  17: core::iter::iterator::Iterator::for_each
  18: xi_core_lib::event_context::EventContext::update_views
  19: xi_core_lib::event_context::EventContext::after_edit
  20: xi_core_lib::event_context::EventContext::do_plugin_cmd
  21: xi_core_lib::tabs::CoreState::plugin_notification
  22: <xi_core_lib::core::WeakXiCore as xi_rpc::Handler>::handle_notification
  23: <xi_rpc::RpcLoop<W>>::mainloop::{{closure}}
  24: crossbeam_utils::thread::scope::{{closure}}
  25: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  26: std::panicking::try::do_call
  27: __rust_maybe_catch_panic
  28: std::panicking::try
  29: std::panic::catch_unwind
  30: crossbeam_utils::thread::scope
  31: <xi_rpc::RpcLoop<W>>::mainloop
  32: xi_core_lib::plugins::start_plugin_process::{{closure}}
[2019-02-08][16:43:35][xi_rpc][ERROR] panic guard hit, closing runloop
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', libcore/result.rs:1009:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::continue_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::unwrap
  10: xi_core_lib::core::XiCore::inner
  11: <xi_core_lib::core::XiCore as xi_rpc::Handler>::handle_notification
  12: <xi_rpc::RpcLoop<W>>::mainloop::{{closure}}
  13: crossbeam_utils::thread::scope::{{closure}}
  14: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  15: std::panicking::try::do_call
  16: __rust_maybe_catch_panic
  17: std::panicking::try
  18: std::panic::catch_unwind
  19: crossbeam_utils::thread::scope
  20: <xi_rpc::RpcLoop<W>>::mainloop
  21: xi_core::main
  22: std::rt::lang_start::{{closure}}
  23: std::panicking::try::do_call
  24: __rust_maybe_catch_panic
  25: std::rt::lang_start_internal
  26: std::rt::lang_start
  27: main
```

Not entirely sure what the problem is.

--- a/rust/core-lib/src/linewrap.rs
+++ b/rust/core-lib/src/linewrap.rs
@@ -246,7 +246,9 @@ impl Lines {
                 new_work.push(head);
                 new_work.push(tail);
             } else {
-                new_work.push(task.translate_neg(iv.size()).translate(new_len));
+                // take task - our edit interval, then translate it (- old_size, + new_size)
+                let tail = task.suffix(iv).translate(new_len).translate_neg(iv.size());
+                new_work.push(tail);
             }
         }
         new_work.retain(|iv| !iv.is_empty());
@@ -1179,5 +1181,16 @@ mod tests {
         assert_eq!(make_ranges(&lines.work), vec![0..20, 30..50]);
         lines.patchup_tasks(10..10, 10);
         assert_eq!(make_ranges(&lines.work), vec![0..30, 40..60]);
+    }
+
+    /// https://github.com/xi-editor/xi-editor/issues/1112
+    #[test]
+    fn patchup_for_edit_before_task() {
+        let mut lines = Lines::default();
+        lines.add_task(0..100);
+        lines.update_tasks_after_wrap(0..30);
+        assert_eq!(make_ranges(&lines.work), vec![30..100]);
+        lines.patchup_tasks(5..90, 80);
+        assert_eq!(make_ranges(&lines.work), vec![85..95]);
     }
 }

